### PR TITLE
Feature create devices after bus creation

### PR DESCRIPTION
**Pull request by akuker**
In preparation of the upcoming banana pi support, I needed to restructure how the program arguments are parsed. Currently, the 'bus' object needs to be created before we can parse the arguments. However, with the bpi updates, I need to have the program arguments as an input to creating the 'bus' object.

Attached is my proposed approach. Basically, it saves off the program args related to device creation, then processes them later (after the bus and associated controllers/executor/etc are created).

### CODE REVIEW

File: cpp/rascsi.cpp [hunk 1]
```diff
@@ -71,6 +71,12 @@ shared_ptr<RascsiResponse> rascsi_response;
 shared_ptr<RascsiExecutor> executor;
 const ProtobufSerializer serializer;
 
+struct optarg_struct{
+	int opt;
+	string value;
+};
+typedef std::vector<optarg_struct> optarg_queue_type;
+
```
Comment (uweseimet) [inline cpp/rascsi.cpp:79]: Please use a standard multimap instead of custom containers here, e.g.:

```
unordered_multimap<int, string> optargs;
```

**Review by uweseimet — CHANGES_REQUESTED** [review]
I'm not familiar with the changes that require creating the devices after the bus. I would expect the devices not to know anything about/depend on the type of bus, and currently the device factory creates them without any information on a bus. Can you please explain?
Maybe this is related to the fact that the device factory knows the controller manager, and this one knows the bus?
The factory currently needs the controller manager only for creating the host services device. I would like to get rid of this dependency, but have not yet had a closer look at alternatives.

**Comment by akuker** [comment]
> I'm not familiar with the changes that require creating the devices after the bus. I would expect the devices not to know anything about/depend on the type of bus, and currently the device factory creates them without any information on a bus. Can you please explain? Maybe this is related to the fact that the device factory knows the controller manager, and this one knows the bus? The factory currently needs the controller manager only for creating the host services device. I would like to get rid of this dependency, but have not yet had a closer look at alternatives.

With the current implementation, the `executor` is used as part of the ParseArgument() routine. The constructor for `executor` uses `controller manager` in its constructor. Controller manager uses `bus` in its constructure. So, with the current architecture, you need to have the `bus` object created in order to run ParseArgument().

With the current implementation of GPIOBUS, this is fine. However, with the upcoming revision, we need to know what type of board we're connected to (fullspec/aibom/standard/gamernium) before we call the GPIOBUS::Init() function. See:
https://github.com/akuker/RASCSI/blob/feature_bpi4/cpp/hal/gpiobus.cpp#L77

Note that this change is NOT NEEDED currently. But, I'd prefer to do any small, benign changes early. If you prefer I can lump this in with the `feature_bpi` branch.

File: cpp/rascsi.cpp (hunk 1 above)
Comment (akuker) [reply cpp/rascsi.cpp:79]: Can do

**Comment by akuker** [comment]
Also note that this moves the log level setting earlier in the start-up process. This can be useful when trying to debug issues during initialization.

**Comment by uweseimet** [comment]
@akuker Looks as if my guess about the manager and its bus argument was quite right. Actually, I don't have a strong opinion on when to merge these changes. I'm fine with doing this early. I intend to resolve the dependency between the controller manager and the bus sooner or later, but this requires also to change the host services and I do not yet know how exactly this change should look like. Otherwise I would already have done it ;-).
So from my perspective, let's merge your changes early, after replacing the vector with the nested struct by a multimap.

File: cpp/rascsi.cpp (hunk 1 above)
Comment (akuker) [reply cpp/rascsi.cpp:75]: I'm wondering if this makes more sense as a FIFO Queue. With the multimap, can we guarantee that the arguments will come out in the same order they went in?

Comment (akuker) [reply cpp/rascsi.cpp:77]: @uweseimet I changed it to a double ended queue (`deque`). let me know if you have any objections!

Comment (uweseimet) [reply cpp/rascsi.cpp:77]: @akuker Ordering would indeed be an issue with the multimap, so we cannot use it here. I am going to approve the PR, but before you merge, please replace "typedef" by "using", as recommended for modern C++. A minor change only.

**Review by uweseimet — APPROVED** [review]

**Comment by uweseimet** [comment]
@akuker I hope I can simplify the parsing again after having a good solution for https://github.com/akuker/RASCSI/issues/959.

File: cpp/rascsi.cpp (hunk 1 above)
Comment (akuker) [reply cpp/rascsi.cpp:77]: Done. Thanks!

**Pull request merged**